Add concept pages for Scanner and Healing (#1192)
minio/docs: source/operations/concepts/healing.rst

.. _minio-concepts-healing:

==============
Object Healing
==============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

What is healing?
----------------

Healing is MinIO's ability to restore an object that has been damaged, corrupted, or partially lost.
The loss can come from random corruptions in the form of 
- :ref:`bit rot <minio-ec-bitrot>`
- the loss of one or more drives due to failure or outages

> - the corruption or loss of one or more shards of an object for other reasons

Reviewer: I wonder if we should order these based on likelihood:

- Drive-level errors or failure
- OS/Filesystem errors or failure 
- :ref:`bit rot <minio-ec-bitrot>` of stored data

Reviewer: and use the 'including but not limited to' preface.

Author: Yes, good ideas on both counts

Healing and Erasure Coding
--------------------------

The ability of MinIO to restore a damaged object relates directly to the following:

- total number of drives in the :ref:`erasure set <minio-erasure-coding>` where the object exists
- :ref:`parity setting <minio-ec-parity>` for the erasure set

>   Parity refers to the number of extra shards MinIO create when writing the object.

Reviewer: I would not reference them as 'extra shards'

Parity refers to dedicated recovery shards MinIO creates when writing an object.
The number of parity shards created per object depends on the erasure parity configuration of the deployment.

Author: Agree with this and updated.

>   Parity refers to the number of extra shards MinIO create when writing the object.

Reviewer 2: ```suggestion
  Parity refers to the number of extra shards MinIO creates when writing the object.
```

>   For example, an erasure set may have eight total drives and use three drives during a write for parity.
>   In that scenario, MinIO splits the object into five parts written to five of the drives.
>   MinIO uses the remaining three drives in the erasure set for parity shards.

Reviewer: This *might* imply that we always use the same drives for parity or data. It's randomized, so maybe

```suggestion
For example, MinIO splits an object into 5 data and 3 parity shards and distributes those shards across 8 drives.
```

Author: 🤔  Interesting thought. Taking your suggestion for the most part.

>   For example, an erasure set may have eight total drives and use three drives during a write for parity.
>   In that scenario, MinIO splits the object into five parts written to five of the drives.
>   MinIO uses the remaining three drives in the erasure set for parity shards.

Reviewer 2: My understanding is hazy, but maybe this is a thing?
```suggestion
  For example, an erasure set may have eight drives in total: five for data shards and three for parity shards.
  In that scenario, MinIO splits the object into five parts, writing each part to five different drives.
  MinIO uses the remaining three drives in the erasure set for parity shards.
```

Author: Ravind had also made a suggestion on this paragraph. I've reworked it.

>   When MinIO needs to provide the object, it looks for the data shards for the object.

Reviewer: We can use any combination of shards to assemble an object, up to the write quorum / data drives (5 shards).

Author: It looks for and uses data shards first. Only if it cannot find the data shards does it go for parity shards. That's why I worded it this way.

>   If any of the data shards are missing or damaged, MinIO uses the parity shards to restore the object.

Reviewer: This goes both ways - we can restore parity shards if missing.

Author: Clarified

  For this scenario, up to three of data shard parts can be lost or damaged and MinIO can still successfully restore and serve the object. 
- number of drives available with intact parts of the object

  The number of drives available with intact data or parity shards of the object must meet or exceed the number of drives used for data shards in the erasure set.
  In the scenario above, five drives with intact shards must be online and available for MinIO to successfully serve the object.

When does MinIO heal an object?
-------------------------------

MinIO has a robust system for healing objects.

> Healing during ``GET`` requests

Reviewer: I think both GET and HEAD (possibly STAT?)

Author: Updated with the info from Harsha.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

MinIO automatically checks the consistency of an object's data shards each time you request an object with a ``GET`` operation.

> If all data shards are found intact, MinIO takes no further action on the object and serves the object from the data shards.

Reviewer: We should be checking all shards, including parity

Author: We don't do anything with parity shards if all data shards are intact.

Author: Clarified the wording, but what I wrote is correct for how it works.
MinIO looks for the data shards and serves the object if all data shards are intact.
Only if any of the data shards are missing or damaged does it do anything with parity shards.

> MinIO automatically checks the consistency of an object's data shards each time you request an object with a ``GET`` operation.
> 
> If all data shards are found intact, MinIO takes no further action on the object and serves the object from the data shards.
> 
> If MinIO finds missing or damaged data shards, MinIO utilizes the available parity shards to heal the object before serving it as part of the ``GET`` operation.
> There **must** be a parity shard available for each lost or damaged data shard.

Reviewer 2: some more ideas that might work?
```suggestion
MinIO automatically checks the consistency of an object's data shards with each ``GET`` operation.

If all the data shards are intact, MinIO serves the object from the data shards without inspecting the corresponding parity shards

If the object has missing or damaged data shards, MinIO uses the available parity shards to heal the object before serving it as part of the ``GET`` operation.
There **must** be a parity shard available for each lost or damaged data shard, otherwise the object cannot be recovered.
```

Author: Taking these. Thanks.

Healing with the object scanner
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

MinIO utilizes an :ref:`object scanner <minio-concepts-scanner>` to perform a number of tasks related to objects.
One of these tasks checks the integrity of objects and, if found damaged or corrupted, heals them.

> On each scanning pass, MinIO selects one out of every 1,024 objects to check.

Reviewer: This might imply we check 1 object per pass/cycle. Not sure how best to address that.

"MinIO checks a portion of the total namespace on each pass, using a filter ratio to determine which objects to check or skip."

Author: Tweaked your suggestion, but running with it.

MinIO chooses the objects based on a hash of the object name.

> If any object is found to have lost data shards, MinIO heals the object from available parity shards.

Reviewer: Same, we check for any missing shards

Author: we don't, not if all data shards are intact by the scanner.

Author: Clarifying how this works internally to confirm my understanding.

Reviewer 3: I don't understand what the disagreement is about.

Reviewer: @klauspost my vague understanding is that we check all shards, data or parity, on scanner hit to ensure that the object is healthy. The above might imply that we do not check for parity shard health.

Reviewer 3: Ah. I see. Yeah ` lost data shards` -> ` lost shards` would be more accurate.

Author: Changed!

> MinIO utilizes an :ref:`object scanner <minio-concepts-scanner>` to perform a number of tasks related to objects.
> One of these tasks checks the integrity of objects and, if found damaged or corrupted, heals them.
> 
> On each scanning pass, MinIO selects one out of every 1,024 objects to check.
> MinIO chooses the objects based on a hash of the object name.
> 
> If any object is found to have lost data shards, MinIO heals the object from available parity shards.
> By default, MinIO does *not* check for :term:`bit rot` corruption using the scanner, as it can be an expensive operation to perform while bit rot is a low risk across multiple disks.

Reviewer 2: some ideas?
```suggestion
MinIO uses an :ref:`object scanner <minio-concepts-scanner>` to perform a number of tasks related to objects.
One of these tasks checks the integrity of objects and, if found damaged or corrupted, heals them.

On each scanning pass, MinIO uses a hash of the object name to select one out of every 1,024 objects to check.

If any object is found to have lost data shards, MinIO heals the object from available parity shards.
By default, MinIO does *not* check for :term:`bit rot` corruption using the scanner.
This can be an expensive operation to perform and the risk of bit rot across multiple disks is low.
```

Could even add something like `To enable scanner bit rot checking, see [ref]`

Author: 👍

Healing by manual request
~~~~~~~~~~~~~~~~~~~~~~~~~

Administrators can use :mc:`mc admin heal` to initiate a full system healing.
The procedure is very resource intensive and not typically needed.

Consult with MinIO Engineers before manually starting a healing process on a deployment.

Healing metrics
---------------

MinIO provides several `healing metrics <https://min.io/docs/minio/linux/operations/monitoring/metrics-and-alerts.html#healing-metrics>`__ to monitor the status of healing processes on a deployment.

Refer to the :ref:`minio-metrics-and-alerts` for more information on available endpoints and configuration.

.. _minio-concepts-healing-colors:

Healing Output Color Key
------------------------

> Originally, the healing mechanism output a table that used a Green-Yellow-Red-Gray color key to attempt to differentiate the status of objects in healing.

Reviewer 2: ```suggestion
Earlier MinIO versions generated a table with green, yellow, red, and gray colors to indicate the healing status of individual objects.
```

Is there a specific version we could reference here? Does it even matter?

These colors have been deprecated in favor of more detailed :ref:`healing metrics available at the cluster level <minio-metrics-and-alerts-available-metrics>`.

> The following table describes the intent of each of the deprecated color keys.

Reviewer 2: more ideas
```suggestion
The following table describes the meaning of the now deprecated status colors:
```

Object Healing
==============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

What is healing?
----------------

Healing is MinIO's ability to restore an object that has been damaged, corrupted, or partially lost.
The loss can come from multiple types of corruptions or loss, such as but not limited to:
- drive-level errors or failure
- OS or filesystem errors or failure
- :ref:`bit rot <minio-ec-bitrot>`

Healing and Erasure Coding
--------------------------

The ability of MinIO to restore a damaged object relates directly to the following:

- total number of drives in the :ref:`erasure set <minio-erasure-coding>` where the object exists
- number of drives available with intact parts of the object
- :ref:`parity setting <minio-ec-parity>` for the erasure set

  :term:`Parity` refers to the number of dedicated recovery shards MinIO creates when writing the object.
  For example, an erasure set may have eight total drives and use three drives during a write for parity.
  In this scenario, MinIO splits an object into 5 data shards and create 3 parity shards.
  MinIO distributes these eight shards across the drives in the erasure set.

>   No one drive contains only parity shards or only data shards as a protection.

Reviewer 3: ```suggestion
  No one drive contains only parity shards or only data shards.
  Instead it is randomized per object to distribute reads evenly across disks.
```

Author: Mildly tweaked from the suggestion, but updated.

Refer to the :ref:`minio-metrics-and-alerts` for more information on available endpoints and configuration.

.. _minio-concepts-healing-colors:

Healing Output Color Key
------------------------

Earlier MinIO versions generated a table with green, yellow, red, and gray colors to indicate the healing status of individual objects.
These colors have been deprecated in favor of more detailed :ref:`healing metrics available at the cluster level <minio-metrics-and-alerts-available-metrics>`.

The following table describes the meaning of the now-deprecated status colors.

.. list-table::
   :widths: 25 75
   :width: 100%

   * - **Green**
     - *Healthy*, the object has all data and parity shards available as required to serve the object
 
   * - **Yellow** 
     - *Healing*, the object is still in the process of healing, and there are sufficient data or parity shards available to complete the healing

   * - **Red** 
     - *Unhealthy*, the object has lost one or more shards and requires healing

   * - **Grey** 
     -  *Unrecoverable*, the object has lost too many data and/or parity shards and cannot be healed or recovered

> 

Reviewer 3: Does all of this matter if it is deprecated?

Reviewer: I think it comes up often enough in support cases, and we have some UI/UX in Console/Enterprise Console that also uses color coding.

Author: As @ravindk89 says, this is mainly for those times when someone comes across it in an older version or in the Console and wants to know what the colors mean. This still happens.

.. _minio-concepts-healing:

==============
Object Healing
==============

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

What is healing?
----------------

Healing is MinIO's ability to restore an object that has been damaged, corrupted, or partially lost.

> The loss can come from multiple types of corruptions or loss, such as but not limited to:

Reviewer 2: Blank line to make list below work correctly
```suggestion
The loss can come from multiple types of corruptions or loss, such as but not limited to:

```
